src/app_builder.rs

```diff
@@ -1,69 +1,241 @@
-#[cfg(feature = "startup-stages")]
-use crate::startup_stage;
-use crate::{app::App, plugin::Plugin, stage};
+use crate::{TrackedValue, app::App, plugin::Plugin, tracked_unique::reset_tracked_unique, type_names::TypeNames};
 use shipyard::*;
 use std::{
     any::{type_name, TypeId},
+    borrow::Cow,
     collections::hash_map::Entry,
     collections::HashMap,
+    rc::Rc,
 };
 use tracing::*;
 
 mod plugin_id;
-mod workloads;
 use plugin_id::PluginId;
-use workloads::Workloads;
+
+/// Name of app stage responsible for doing most app logic. Systems should be registered here by default.
+pub const DEFAULT_STAGE: &str = "default";
+
+#[derive(Clone, Debug)]
+pub struct PluginAssociated {
+    pub plugin: PluginId,
+    pub reason: &'static str,
+}
+
+pub(crate) type PluginsAssociatedMap = TypeIdBuckets<PluginAssociated>;
+
+pub(crate) struct TypeIdBuckets<T> {
+    pub(crate) name: &'static str,
+    pub(crate) track_type_names: TypeNames,
+    pub(crate) type_plugins_lookup: HashMap<TypeId, Vec<T>>,
+}
+
+impl<T: std::fmt::Debug + Clone> std::fmt::Debug for TypeIdBuckets<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_map()
+            .entry(&"association", &self.name)
+            .entries(
+                self.entries()
+                    .into_iter()
+                    .map(|((_, name), associated)| (name, associated)),
+            )
+            .finish()
+    }
+}
+
+pub enum AssociateResult {
+    First,
+    Nth(usize),
+}
```

Comment:
This could be made smaller with:

```suggestion
pub struct AssociateResult {
    previous_assocs: usize,
}
```

And have`fn is_first(&self) -> bool { self.previous_assocs == 0 }`.

Reply:
Yeah, true. The thought crossed my mind. I'll apply

Reply:
I originally was thinking of just making `{ nth: usize, }`